spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/filter/headers/observation/ObservedRequestHttpHeadersFilter.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2013-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gateway.filter.headers.observation;
+
+import io.micrometer.common.lang.Nullable;
+import io.micrometer.observation.Observation;
+import io.micrometer.observation.ObservationRegistry;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.cloud.gateway.filter.headers.HttpHeadersFilter;
+import org.springframework.http.HttpHeaders;
+import org.springframework.web.server.ServerWebExchange;
+
+/**
+ * Observation of {@link HttpHeadersFilter} for a request. It will start an
+ * {@link Observation} when the requests are being filtered. The {@link Observation} will
+ * be stopped when a response is being parsed via
+ * {@link ObservedResponseHttpHeadersFilter}.
+ *
+ * @author Marcin Grzejszczak
+ * @since 4.0.0
+ */
+public class ObservedRequestHttpHeadersFilter implements HttpHeadersFilter {
+
+	private static final Log log = LogFactory.getLog(ObservedRequestHttpHeadersFilter.class);
+
+	static final String CHILD_OBSERVATION = "gateway.observation";
+
+	static final String CHILD_OBSERVATION_CONTEXT = "gateway.observation.context";
+
+	private final ObservationRegistry observationRegistry;
+
+	@Nullable
+	private final GatewayObservationConvention customGatewayObservationConvention;
+
+	public ObservedRequestHttpHeadersFilter(ObservationRegistry observationRegistry) {
+		this(observationRegistry, null);
+	}
+
+	public ObservedRequestHttpHeadersFilter(ObservationRegistry observationRegistry,
+			@Nullable GatewayObservationConvention customGatewayObservationConvention) {
+		this.observationRegistry = observationRegistry;
+		this.customGatewayObservationConvention = customGatewayObservationConvention;
+	}
+
+	@Override
+	public HttpHeaders filter(HttpHeaders input, ServerWebExchange exchange) {
+		HttpHeaders newHeaders = new HttpHeaders();
+		newHeaders.putAll(input);
+		if (log.isDebugEnabled()) {
+			log.debug("Will instrument the HTTP request headers " + newHeaders);
+		}
+		Observation parentObservation = exchange.getAttribute("micrometer.observation");
```

Comment:
Where does `"micrometer.observation"` come from?

Reply:
It comes from https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-observation/src/main/java/io/micrometer/observation/contextpropagation/ObservationThreadLocalAccessor.java#L37 - we can make a direct reference to that